src/ethereum_test_tools/common/types.py

```diff
@@ -489,13 +503,71 @@ def to_kwargs_dict(account: "Dict | Account | None") -> Dict:
         return cls(**kwargs)
 
 
+class Sender(Address):
+    """
+    Address of the sender of a transaction.
+    """
+
+    key: Hash | None
+    nonce: Number
+
+    def __new__(
+        cls,
+        address: "FixedSizeBytesConvertible | Address | Sender | None" = None,
+        *,
+        key: FixedSizeBytesConvertible | None = None,
+        nonce: NumberConvertible = 0,
+    ):
+        """
+        Init the sender.
+        """
+        if address is None:
+            if key is None:
+                raise ValueError("impossible to initialize sender without address")
+            private_key = PrivateKey(Hash(key))
+            public_key = private_key.public_key
+            address = Address(keccak256(public_key.format(compressed=False)[1:])[32 - 20 :])
+        elif isinstance(address, Sender):
+            return address
+        instance = super(Sender, cls).__new__(cls, address)
+        instance.key = Hash(key) if key is not None else None
+        instance.nonce = Number(nonce)
+        return instance
+
+    def copy(self) -> "Sender":
+        """
+        Returns a copy of the sender.
+        """
+        return Sender(Address(self), key=self.key, nonce=self.nonce)
+
+
+MAX_SENDERS = 50
+SENDERS_ITER = iter(
+    Sender(key=TestPrivateKey + i if i != 1 else TestPrivateKey2, nonce=0) for i in count(0)
+)
+SENDERS = [next(SENDERS_ITER) for _ in range(MAX_SENDERS)]
```

Comment:
Couldn't we replace this module-scope variable with an `Alloc` class method along the lines of the following?

```python
    @staticmethod
    def generate_eoa() -> Generator[Sender, None, None]:
        """
        Generate an iterable of EOAs.
        """
        for i in count(0):
            key = TestPrivateKey2 if i == 1 else TestPrivateKey + i
            yield Sender(key=key, nonce=0)
```  

This avoids generating lists of objects upon import and stops people importing SENDERS directly (which my be undesirable and result in side-effects).

If you like this, we could even make `TestPrivateKey` a constructor argument to `Alloc`, so we can generate tests which don't have conflicting EOAs (or senders). This would help us avoid address collisions for batching tests within one genesis.

Reply:
Originally something like this was the idea, the problem is that the method returns a new generator each time, and without being able to save it somewhere between calls, the generator will reset every time.

I think sub-classing is basically our only option until pydantic allows extra fields in the `RootModel`, unless I'm missing something.

Reply:
Added a cached function called `eoa_by_index` which does not require initialization on module import but still reuses previously calculated EOAs.